dracut/30ignition/coreos-persist-hostname.service

```diff
@@ -0,0 +1,25 @@
+# Set the first-boot hostname as the permanent hostname.
+
+[Unit]
+Description=Persist firstboot hostname
+DefaultDependencies=false
+
+# Only trigger when the hostname is NOT A default.
+ConditionHost=!localhost
+ConditionHost=!localhost.localdomain
```

Comment:
Hmm why a separate unit versus doing this as part of the existing logic for `coreos-teardown-initramfs-network.service` ?

Which currently only triggers effectively when the kernel cmdline has a non-default `ip=` - i.e. basically a static IP.

Reply:
The separate unit made the logic simplifier, e.g. `ConditionHost=localhost` and because this also applies to the DHCP situation. For RHCOS we don't want transient names to be used -- the firstboot hostname should be persisted. This ensures that.

Reply:
Are we really sure this applies to DHCP too?  That scopes this code into running in e.g. AWS/GCP etc. as well as the baremetal environments that use DHCP.

Now, I don't think it's at all supported to change the hostnames returned from DHCP for OpenShift/Kubernetes today.  So, it's probably not going to *hurt* to persist it and thereafter ignore what DHCP says, but...no one has been asking us to do that either.

One thing to bear in mind here is that Kubernetes also requires the hostname to be *routable* - which implies it has to be correct in the infrastructure and not *just* the OS.  Hence, source-of-truth is the infrastructure.

The way I'd been thinking of this is that the argument for why this is RHCOS specific is that in the future in FCOS (and ideally RHCOS) with NM-in-initrd, it would parse the kernel arguments and handle this itself, correctly transferring state across the initrd.

Reply:
> Are we really sure this applies to DHCP too?

Better question: do we _want_ to allow changing the host via DHCP for OCP?

Reply:
I think we all agree that doesn't work.  The question is more - does fixing the hostname for DHCP *help* anything either?  

Now, I know for sure that if we had shipped this patch from the start it would have completely avoided the need for the [kubelet waiting on NM](https://github.com/openshift/machine-config-operator/pull/1206) patch.  But, we are on track to ship that patch long before this one.